File: lib/world/simulation.ts
// ... keeping imports (lines 1-4)
import { generateText, tool } from 'ai';
import { z } from 'zod';
import { openrouter, models } from '@/lib/ai/openrouter';
import type { Character, WorldState, WorldEvent, Message, Conversation } from '@/types/world';

type SimulationDepth = 'full' | 'summary' | 'skip';

/**
 * Determine how deeply to simulate interactions between characters.
 */
function determineSimulationDepth(
  timeSinceLastInteraction: number,
  hasUnresolvedPlotPoints: boolean
): SimulationDepth {
  if (timeSinceLastInteraction < 5) return 'skip';
  if (hasUnresolvedPlotPoints || timeSinceLastInteraction > 20) return 'full';
  if (timeSinceLastInteraction > 10) return 'summary';
  return 'skip';
}

/**
 * Group characters by their current location.
 */
function groupCharactersByLocation(characters: Character[]): Map<string, Character[]> {
  const groups = new Map<string, Character[]>();
  for (const char of characters) {
    const existing = groups.get(char.currentLocationClusterId) || [];
    existing.push(char);
    groups.set(char.currentLocationClusterId, existing);
  }
  return groups;
}

/**
 * Generate a summary of what characters did while apart.
 */
async function generateSummary(
  characters: Character[],
  locationName: string,
  timeElapsed: number,
  world: WorldState,
  modelId?: string
): Promise<WorldEvent> {
  const characterNames = characters.map(c => c.name).join(' and ');

  const { text } = await generateText({
    model: openrouter(modelId || models.fast),
    prompt: `Summarize what likely happened between ${characterNames} over ${timeElapsed} time units at ${locationName}.

Characters:
${characters.map(c => `- ${c.name}: ${c.description}${c.goals ? `\n  Goal: ${c.goals}` : ''}`).join('\n')}

Scenario: ${world.scenario.description}

Write a brief 1-2 sentence summary of their interactions. Be specific but concise.`,
  });

  return {
    id: crypto.randomUUID(),
    timestamp: world.time.tick,
    locationClusterId: characters[0].currentLocationClusterId,
    involvedCharacterIds: characters.map(c => c.id),
    description: text.trim(),
    witnessedByIds: characters.map(c => c.id),
    isOffscreen: true,
  };
}

/**
 * Run a full dialogue simulation between characters.
 */
async function runFullSimulation(
  characters: Character[],
  locationName: string,
  timeElapsed: number,
  world: WorldState,
  modelId?: string,
  relevantEvents: WorldEvent[] = []
): Promise<{
  events: WorldEvent[];
  messages: Message[];
  conversation: Omit<Conversation, 'id'>,
  movements: { characterId: string; newLocationId: string }[]
}> {
  const characterNames = characters.map(c => c.name).join(' and ');
  const turnCount = Math.min(Math.ceil(timeElapsed / 2), 8);

  // Get available locations for characters to potentially move to
  const availableLocations = world.locationClusters
    .map(l => l.canonicalName)
    .join(', ');

  const relationshipsText = characters.map(c => {
    const rels = c.relationships.filter(r => characters.some(other => other.id === r.characterId));
    if (rels.length === 0) return '';
    return `- ${c.name}'s Relationships:\n${rels.map(r =>
      `  * With ${characters.find(ch => ch.id === r.characterId)?.name}: ${r.description} (Sentiment: ${r.sentiment})`
    ).join('\n')}`;
  }).filter(t => t).join('\n\n');

  const historyText = relevantEvents.length > 0
    ? `SHARED HISTORY (Recent events they know about):\n${relevantEvents.map(e => `- [${e.timestamp}] ${e.description}`).join('\n')}`
    : 'SHARED HISTORY: None recently.';

  const systemPrompt = `You are simulating a conversation between ${characterNames} at ${locationName}.

Characters:
${characters.map(c => `- ${c.name}: ${c.description}${c.goals ? `\n  Goal: ${c.goals}` : ''}`).join('\n')}

${relationshipsText}

${historyText}

Scenario: ${world.scenario.description}
Time: ${world.time.narrativeTime}
Available Locations (for movement): ${availableLocations}

Write a natural dialogue between these characters. Each character should stay in character.
Format each line as: CHARACTER_NAME: "dialogue"
Include brief action descriptions in *asterisks* when appropriate.
If characters decide to go somewhere else, they should express it in dialogue.

Generate approximately ${turnCount} exchanges.`;

  const { text } = await generateText({
    model: openrouter(modelId || models.mainConversation),
    prompt: systemPrompt,
  });

  // Parse the dialogue into messages
  const messages: Message[] = [];
  const lines = text.split('\n').filter(line => line.trim());

  for (const line of lines) {
    const match = line.match(/^([A-Za-z]+):\s*(.+)$/);
    if (match) {
      const [, name, content] = match;
      const speaker = characters.find(c =>
        c.name.toLowerCase() === name.toLowerCase()
      );

      messages.push({
        id: crypto.randomUUID(),
        conversationId: '',
        role: 'assistant',
        content: content.replace(/^["']|["']$/g, '').trim(),
        timestamp: world.time.tick,
        speakerId: speaker?.id,
      });
    }
  }

  // Extract key events and movements using tool calling
  const result = await generateText({
    model: openrouter(modelId || models.fast),
    tools: {
      reportSimulation: tool({
        description: 'Report events and movements from the conversation',
        inputSchema: z.object({
          events: z.array(z.object({
            description: z.string().describe('Brief description of what happened'),
            isSignificant: z.boolean().describe('Whether this is plot-relevant'),
          })),
          movements: z.array(z.object({
            characterName: z.string(),
            destination: z.string().describe('Name of the location they are going to'),
          })).optional(),
        }),
      }),
    },
    toolChoice: 'required',
    prompt: `Analyze this conversation and extract significant events and any character movements:

${text}

List any important events (agreements made, information shared, conflicts).
If any character EXPLICITLY decides to leave for another location, report it in movements. Matches must be from: ${availableLocations}`,
  });

  let extractedEvents: { description: string; isSignificant: boolean }[] = [];
  let extractedMovements: { characterName: string; destination: string }[] = [];

  const toolCall = result.toolCalls[0] as any;
  if (toolCall && toolCall.toolName === 'reportSimulation') {
    extractedEvents = toolCall.input.events;
    extractedMovements = toolCall.input.movements || [];
  }

  const events: WorldEvent[] = extractedEvents
    .filter(e => e.isSignificant)
    .map(e => ({
      id: crypto.randomUUID(),
      timestamp: world.time.tick,
      locationClusterId: characters[0].currentLocationClusterId,
      involvedCharacterIds: characters.map(c => c.id),
      description: e.description,
      witnessedByIds: characters.map(c => c.id),
      isOffscreen: true,
    }));

  const conversation: Omit<Conversation, 'id'> = {
    type: 'offscreen',
    locationClusterId: characters[0].currentLocationClusterId,
    participantIds: characters.map(c => c.id),
    messages,
    isActive: true,
  };

  // Resolve movements to IDs
  const movements: { characterId: string; newLocationId: string }[] = [];
  for (const move of extractedMovements) {
    const char = characters.find(c => c.name.toLowerCase() === move.characterName.toLowerCase());
    const loc = world.locationClusters.find(l =>
      l.canonicalName.toLowerCase().includes(move.destination.toLowerCase()) ||
      move.destination.toLowerCase().includes(l.canonicalName.toLowerCase())
    );

    if (char && loc && loc.id !== char.currentLocationClusterId) {
      movements.push({
        characterId: char.id,
        newLocationId: loc.id
      });
    }
  }

  return { events, messages, conversation, movements };
}

/**
 * Simulate off-screen interactions between characters.
 */
export async function simulateOffscreen(
  world: WorldState,
  playerLocationClusterId: string,
  timeSinceLastSimulation: number,
  modelId?: string,
  relevantEvents: WorldEvent[] = []
): Promise<{
  events: WorldEvent[];
  conversations: Omit<Conversation, 'id'>[];
  characterUpdates: { characterId: string; newLocationId: string }[];
}> {
  // Get non-player characters not at player's location
  const absentCharacters = world.characters.filter(
    c => !c.isPlayer &&
      c.isDiscovered &&
      c.currentLocationClusterId !== playerLocationClusterId
  );

  if (absentCharacters.length < 2) {
    return { events: [], conversations: [], characterUpdates: [] };
  }

  // Group by location
  const byLocation = groupCharactersByLocation(absentCharacters);

  // OPTIMIZATION: Process locations in parallel using Promise.all
  // This significantly reduces total simulation time compared to sequential processing
  const promises = Array.from(byLocation).map(async ([locationId, chars]) => {
    if (chars.length < 2) return null;

    const location = world.locationClusters.find(c => c.id === locationId);
    const locationName = location?.canonicalName ?? 'an unknown location';

    const depth = determineSimulationDepth(timeSinceLastSimulation, false);

    if (depth === 'skip') return null;

    if (depth === 'full') {
      const { events, conversation, movements } = await runFullSimulation(
        chars,
        locationName,
        timeSinceLastSimulation,
        world,
        modelId,
        relevantEvents
      );
      return { type: 'full' as const, events, conversation, movements };

    } else if (depth === 'summary') {
      const event = await generateSummary(chars, locationName, timeSinceLastSimulation, world, modelId);
      return { type: 'summary' as const, event };
    }
    return null;
  });

  const results = await Promise.all(promises);

  const allEvents: WorldEvent[] = [];
  const allConversations: Omit<Conversation, 'id'>[] = [];
  const allUpdates: { characterId: string; newLocationId: string }[] = [];

  for (const res of results) {
    if (!res) continue;
    if (res.type === 'full') {
      allEvents.push(...res.events);
      allConversations.push(res.conversation);
      allUpdates.push(...res.movements);
    } else if (res.type === 'summary') {
      allEvents.push(res.event);
    }
  }

  return { events: allEvents, conversations: allConversations, characterUpdates: allUpdates };
}

File: internal/world/simulation.go
package world

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"emergent/internal/ai"
	"emergent/internal/models"

	"github.com/google/uuid"
)

// SimulationResult from off-screen simulation
type SimulationResult struct {
	Events           []models.WorldEvent   `json:"events"`
	Conversations    []models.Conversation `json:"conversations"`
	CharacterUpdates []CharacterUpdate     `json:"characterUpdates"`
}

// CharacterUpdate is a character movement
type CharacterUpdate struct {
	CharacterID   string `json:"characterId"`
	NewLocationID string `json:"newLocationId"`
}

type simulationDepth string

const (
	depthFull    simulationDepth = "full"
	depthSummary simulationDepth = "summary"
	depthSkip    simulationDepth = "skip"
)

func determineSimulationDepth(timeSinceLastInteraction int, hasUnresolvedPlotPoints bool) simulationDepth {
	if timeSinceLastInteraction < 5 {
		return depthSkip
	}
	if hasUnresolvedPlotPoints || timeSinceLastInteraction > 20 {
		return depthFull
	}
	if timeSinceLastInteraction > 10 {
		return depthSummary
	}
	return depthSkip
}

func groupCharactersByLocation(characters []models.Character) map[string][]models.Character {
	groups := make(map[string][]models.Character)
	for _, c := range characters {
		groups[c.CurrentLocationClusterID] = append(groups[c.CurrentLocationClusterID], c)
	}
	return groups
}

func generateSummary(ctx context.Context, characters []models.Character, locationName string, timeElapsed int, world *models.WorldState, modelID string) (*models.WorldEvent, error) {
	var names []string
	for _, c := range characters {
		names = append(names, c.Name)
	}
	characterNames := strings.Join(names, " and ")

	var charDescriptions strings.Builder
	for _, c := range characters {
		fmt.Fprintf(&charDescriptions, "- %s: %s", c.Name, c.Description)
		if c.Goals != "" {
			fmt.Fprintf(&charDescriptions, "\n  Goal: %s", c.Goals)
		}
		charDescriptions.WriteString("\n")
	}

	messages := []ai.ChatMessage{
		{
			Role: "user",
			Content: fmt.Sprintf(`Summarize what likely happened between %s over %d time units at %s.

Characters:
%s
Scenario: %s

Write a brief 1-2 sentence summary of their interactions. Be specific but concise.`,
				characterNames, timeElapsed, locationName, charDescriptions.String(), world.Scenario.Description),
		},
	}

	resp, err := ai.GenerateText(ctx, modelID, messages, nil, nil)
	if err != nil {
		return nil, err
	}

	text := ""
	if len(resp.Choices) > 0 {
		text = strings.TrimSpace(resp.Choices[0].Message.Content)
	}

	var charIDs []string
	for _, c := range characters {
		charIDs = append(charIDs, c.ID)
	}

	return &models.WorldEvent{
		ID:                   uuid.New().String(),
		Timestamp:            world.Time.Tick,
		LocationClusterID:    characters[0].CurrentLocationClusterID,
		InvolvedCharacterIDs: charIDs,
		Description:          text,
		WitnessedByIDs:       charIDs,
		IsOffscreen:          true,
	}, nil
}

var dialogueLineRe = regexp.MustCompile(`^([A-Za-z][A-Za-z ]+):\s*(.+)$`)

func runFullSimulation(ctx context.Context, characters []models.Character, locationName string, timeElapsed int, world *models.WorldState, modelID string) (*SimulationResult, error) {
	var names []string
	for _, c := range characters {
		names = append(names, c.Name)
	}
	characterNames := strings.Join(names, " and ")
	turnCount := max(min(timeElapsed/2, 8), 1)

	var locNames []string
	for _, l := range world.LocationClusters {
		locNames = append(locNames, l.CanonicalName)
	}
	availableLocations := strings.Join(locNames, ", ")

	var charDescriptions strings.Builder
	for _, c := range characters {
		fmt.Fprintf(&charDescriptions, "- %s: %s", c.Name, c.Description)
		if c.Goals != "" {
			fmt.Fprintf(&charDescriptions, "\n  Goal: %s", c.Goals)
		}
		charDescriptions.WriteString("\n")
	}

	systemPrompt := fmt.Sprintf(`You are simulating a conversation between %s at %s.

Characters:
%s
Scenario: %s
Time: %s
Available Locations (for movement): %s

Write a natural dialogue between these characters. Each character should stay in character.
Format each line as: CHARACTER_NAME: "dialogue"
Include brief action descriptions in *asterisks* when appropriate.
If characters decide to go somewhere else, they should express it in dialogue.

Generate approximately %d exchanges.`,
		characterNames, locationName, charDescriptions.String(), world.Scenario.Description,
		world.Time.NarrativeTime, availableLocations, turnCount)

	messages := []ai.ChatMessage{
		{Role: "user", Content: systemPrompt},
	}

	resp, err := ai.GenerateText(ctx, modelID, messages, nil, nil)
	if err != nil {
		return &SimulationResult{}, err
	}

	text := ""
	if len(resp.Choices) > 0 {
		text = resp.Choices[0].Message.Content
	}

	// Parse dialogue into messages
	var convMessages []models.Message
	lines := strings.SplitSeq(text, "\n")
	for line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		matches := dialogueLineRe.FindStringSubmatch(line)
		if matches == nil {
			continue
		}
		speakerName := matches[1]
		content := strings.Trim(strings.TrimSpace(matches[2]), `"'`)

		var speakerID string
		for _, c := range characters {
			if strings.EqualFold(c.Name, speakerName) {
				speakerID = c.ID
				break
			}
		}

		convMessages = append(convMessages, models.Message{
			ID:             uuid.New().String(),
			ConversationID: "",
			Role:           "assistant",
			Content:        content,
			Timestamp:      world.Time.Tick,
			SpeakerID:      speakerID,
		})
	}

	// Extract events and movements using tool calling
	tools := []ai.Tool{
		{
			Type: "function",
			Function: ai.ToolFunction{
				Name:        "reportSimulation",
				Description: "Report events and movements from the conversation",
				Parameters: map[string]any{
					"type": "object",
					"properties": map[string]any{
						"events": map[string]any{
							"type": "array",
							"items": map[string]any{
								"type": "object",
								"properties": map[string]any{
									"description":   map[string]any{"type": "string"},
									"isSignificant": map[string]any{"type": "boolean"},
								},
								"required": []string{"description", "isSignificant"},
							},
						},
						"movements": map[string]any{
							"type": "array",
							"items": map[string]any{
								"type": "object",
								"properties": map[string]any{
									"characterName": map[string]any{"type": "string"},
									"destination":   map[string]any{"type": "string"},
								},
								"required": []string{"characterName", "destination"},
							},
						},
					},
					"required": []string{"events"},
				},
			},
		},
	}

	extractMessages := []ai.ChatMessage{
		{
			Role: "user",
			Content: fmt.Sprintf(`Analyze this conversation and extract significant events and any character movements:

%s

List any important events (agreements made, information shared, conflicts).
If any character EXPLICITLY decides to leave for another location, report it in movements. Matches must be from: %s`,
				text, availableLocations),
		},
	}

	extractResp, err := ai.GenerateText(ctx, modelID, extractMessages, tools, "required")

	var extractedEvents []struct {
		Description   string `json:"description"`
		IsSignificant bool   `json:"isSignificant"`
	}
	var extractedMovements []struct {
		CharacterName string `json:"characterName"`
		Destination   string `json:"destination"`
	}

	if err == nil && len(extractResp.Choices) > 0 && len(extractResp.Choices[0].Message.ToolCalls) > 0 {
		tc := extractResp.Choices[0].Message.ToolCalls[0]
		if tc.Function.Name == "reportSimulation" {
			var args struct {
				Events []struct {
					Description   string `json:"description"`
					IsSignificant bool   `json:"isSignificant"`
				} `json:"events"`
				Movements []struct {
					CharacterName string `json:"characterName"`
					Destination   string `json:"destination"`
				} `json:"movements"`
			}
			if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err == nil {
				extractedEvents = args.Events
				extractedMovements = args.Movements
			}
		}
	}

	var charIDs []string
	for _, c := range characters {
		charIDs = append(charIDs, c.ID)
	}

	var events []models.WorldEvent
	for _, e := range extractedEvents {
		if !e.IsSignificant {
			continue
		}
		events = append(events, models.WorldEvent{
			ID:                   uuid.New().String(),
			Timestamp:            world.Time.Tick,
			LocationClusterID:    characters[0].CurrentLocationClusterID,
			InvolvedCharacterIDs: charIDs,
			Description:          e.Description,
			WitnessedByIDs:       charIDs,
			IsOffscreen:          true,
		})
	}

	convID := uuid.New().String()
	conversation := models.Conversation{
		ID:                convID,
		Type:              "offscreen",
		LocationClusterID: characters[0].CurrentLocationClusterID,
		ParticipantIDs:    charIDs,
		Messages:          convMessages,
		IsActive:          true,
	}

	// Resolve movements
	var movements []CharacterUpdate
	for _, move := range extractedMovements {
		var charID string
		for _, c := range characters {
			if strings.EqualFold(c.Name, move.CharacterName) {
				charID = c.ID
				break
			}
		}
		if charID == "" {
			continue
		}

		var locID string
		// Prefer exact match
		for _, l := range world.LocationClusters {
			if strings.EqualFold(l.CanonicalName, move.Destination) {
				locID = l.ID
				break
			}
		}
		// Fall back to substring match
		if locID == "" {
			destLower := strings.ToLower(move.Destination)
			for _, l := range world.LocationClusters {
				nameLower := strings.ToLower(l.CanonicalName)
				if strings.Contains(nameLower, destLower) || strings.Contains(destLower, nameLower) {
					locID = l.ID
					break
				}
			}
		}
		if locID == "" {
			continue
		}

		// Don't move if already there
		for _, c := range characters {
			if c.ID == charID && c.CurrentLocationClusterID == locID {
				locID = ""
				break
			}
		}
		if locID != "" {
			movements = append(movements, CharacterUpdate{
				CharacterID:   charID,
				NewLocationID: locID,
			})
		}
	}

	return &SimulationResult{
		Events:           events,
		Conversations:    []models.Conversation{conversation},
		CharacterUpdates: movements,
	}, nil
}

// SimulateOffscreen runs off-screen simulation for non-player characters
func SimulateOffscreen(ctx context.Context, world *models.WorldState, playerLocationClusterID string, timeSinceLastSimulation int, modelID string) (*SimulationResult, error) {
	// Get absent, discovered, non-player characters
	var absentCharacters []models.Character
	for _, c := range world.Characters {
		if !c.IsPlayer && c.IsDiscovered && c.CurrentLocationClusterID != playerLocationClusterID {
			absentCharacters = append(absentCharacters, c)
		}
	}

	if len(absentCharacters) < 2 {
		return &SimulationResult{}, nil
	}

	byLocation := groupCharactersByLocation(absentCharacters)

	result := &SimulationResult{}

	for locationID, chars := range byLocation {
		if len(chars) < 2 {
			continue
		}

		locationName := "an unknown location"
		for _, l := range world.LocationClusters {
			if l.ID == locationID {
				locationName = l.CanonicalName
				break
			}
		}

		depth := determineSimulationDepth(timeSinceLastSimulation, false)
		if depth == depthSkip {
			continue
		}

		if depth == depthFull {
			simResult, err := runFullSimulation(ctx, chars, locationName, timeSinceLastSimulation, world, modelID)
			if err != nil {
				continue
			}
			result.Events = append(result.Events, simResult.Events...)
			result.Conversations = append(result.Conversations, simResult.Conversations...)
			result.CharacterUpdates = append(result.CharacterUpdates, simResult.CharacterUpdates...)
		} else if depth == depthSummary {
			event, err := generateSummary(ctx, chars, locationName, timeSinceLastSimulation, world, modelID)
			if err != nil {
				continue
			}
			result.Events = append(result.Events, *event)
		}
	}

	return result, nil
}
